### PR TITLE
[ws-rollout] Add prometheus init check

### PR DESCRIPTION
**Pull request by Pothulapati**
## Description

When given a non-usable `--prometheus-url`, We start the rollout without verifying if the prometheus is reachable or not. This is a problem as we will be unable to get the metrics from prometheus and hence the rollout will be reverted later causing unnecessary time waste.

This can be prevented by performing a simple check to see if the prometheus is reachable or not. `up` query is used instead of key metrics as we can't be sure of their existence.

Signed-off-by: Tarun Pothulapati <tarun@gitpod.io>

## Related Issue(s)

Fixes #

## How to test

Testing by passing a unreachable prometheus URL

```bash
gitpod /workspace/gitpod/components/workspace-rollout-job (tar/ws-rollout-prom-init) $ go run . --new-cluster g2c4ec3d5cb --old-cluster gfd9876cfe2 --prometheus-url http://localhos:9090
INFO[0000] Starting workspace-rollout-job               
FATA[0001] init: prometheus is not reachable             error="Post \"http://localhos:9090/api/v1/query\": dial tcp: lookup localhos on 1.1.1.1:53: no such host"
exit status 1
```

## Release Notes

```release-note
[ws-rollout] Add prometheus init check
```

## Build Options:

- [ ] /werft with-github-actions
      Experimental feature to run the build with GitHub Actions (and not in Werft).
- [ ] leeway-no-cache
      leeway-target=components:all-ci
- [ ] /werft no-test
      Run Leeway with `--dont-test`
- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [ ] /werft with-preview
- [ ] /werft with-large-vm
- [ ] /werft with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`

### CODE REVIEW

File: components/workspace-rollout-job/cmd/root.go [hunk 1]
```diff
@@ -99,6 +99,13 @@ var rootCmd = &cobra.Command{
 			return err
 		}
 
+		// Check if prometheus is reachable
+		err = analysis.CheckPrometheusReachable(ctx, conf.prometheusURL)
+		if err != nil {
+			log.WithError(err).Fatal("init: prometheus is not reachable")
+			return err
```
Comment (WVerlaek) [inline components/workspace-rollout-job/cmd/root.go:106]: could remove the error return here, `log.Fatal` will already exit the process

File: components/workspace-rollout-job/pkg/analysis/analyzer.go [hunk 2]
```diff
@@ -21,3 +25,24 @@ type Analyzer interface {
 	// repeatedly to determine whether to move forward on the rollout or not.
 	MoveForward(ctx context.Context, clusterName string) (Decision, error)
 }
+
+func CheckPrometheusReachable(ctx context.Context, prometheusURL string) error {
+	if prometheusURL == "" {
+		return nil
+	}
```
Comment (WVerlaek) [inline components/workspace-rollout-job/pkg/analysis/analyzer.go:32]: What about moving this function to `WorkspaceKeyMetricsAnalyzer`, e.g. `WorkspaceKeyMetricsAnalyzer.CheckPrometheusReachable()`, and then use the `api` client of the analyzer to check if it's reachable. You'd only need the `v1Client.Query(ctx, "up", time.Now())` part then, deduplicates the construction of the API clients

**Review by WVerlaek — COMMENTED** [review]
Nice - have some non-blocking suggestions

**Review by WVerlaek — APPROVED** [review]

**Comment by kylos101** [comment]
@gitpod-io/engineering-delivery-operations-experience could we ask for your help with a review? It'd be great to test drive ws-rollout next week during the traffic shift.

**Comment by ArthurSens** [comment]
@kylos101 since I was the one reviewing tarun's work before, I can say that the PR looks good. @WVerlaek's review makes sense though, but I'm not sure @Pothulapati can push changes to the branch since he is not part of the org anymore (nor if he is willing to do so)

@corneliusludmann could you give the ✅ just so the change gets merged and you'll apply the fix in another PR?

**Review by nandajavarma — APPROVED** [review]

**Comment by kylos101** [comment]
/unhold

We'll handle the suggestions in a follow-on PR, if needed, thank you @WVerlaek  for the comprehensive review! 

Also, @nandajavarma appreciate you unblocking this!

**Pull request merged**